File: AdjoinableMPIexamples/Tapenade_F77/IsendRecvWaitTwoWay/programAMPI_d.f
c   DESIRED TAPENADE TANGENT USING THE AMPI LIBRARY
c Simple Isend+Recv+Wait || Recv+Isend+Wait

      SUBROUTINE head_d(x,xd,y,yd)
      implicit none
      include 'ampi/ampif.h'
      DOUBLE PRECISION xd,yd
      DOUBLE PRECISION x,y
      INTEGER world_rank, err_code, req
      DOUBLE PRECISION locald
      DOUBLE PRECISION local

      call AMPI_Comm_rank(MPI_COMM_WORLD, world_rank, err_code)
      IF (world_rank.EQ.0) THEN
         xd=xd*2
         x=x*2
         call TLS_AMPI_ISend(x, xd, 1,
     +        AMPI_ADOUBLE_PRECISION, AMPI_ADOUBLE_PRECISION, 1, 0,
     +        AMPI_TO_RECV, MPI_COMM_WORLD, req, err_code)
         call TLS_AMPI_Recv(y, yd, 1,
     +        AMPI_ADOUBLE_PRECISION, AMPI_ADOUBLE_PRECISION, 1, 0,
     +        AMPI_FROM_ISEND_WAIT, MPI_COMM_WORLD,
     +        MPI_STATUS_IGNORE, err_code)
         call TLS_AMPI_Wait(req,MPI_STATUS_IGNORE, err_code)
         yd=yd*3
         y=y*3
      ELSE IF (world_rank.EQ.1) THEN
         call TLS_AMPI_Recv(local, locald, 1,
     +        AMPI_ADOUBLE_PRECISION, AMPI_ADOUBLE_PRECISION, 0, 0,
     +        AMPI_FROM_ISEND_WAIT, MPI_COMM_WORLD,
     +        MPI_STATUS_IGNORE, err_code)
         locald = COS(local)*locald
         local = SIN(local)
         call TLS_AMPI_ISend(local, locald, 1,
     +        AMPI_ADOUBLE_PRECISION, AMPI_ADOUBLE_PRECISION, 0, 0,
     +        AMPI_TO_RECV, MPI_COMM_WORLD, req, err_code)
         call TLS_AMPI_Wait(req,MPI_STATUS_IGNORE, err_code)
      ENDIF
      END

      PROGRAM main
      implicit none
      include 'ampi/ampif.h'
      INTEGER world_rank, err_code
      DOUBLE PRECISION xd,yd
      DOUBLE PRECISION x,y

      call AMPI_Init_NT(err_code)
      call AMPI_Comm_rank(MPI_COMM_WORLD, world_rank, err_code)
      IF (world_rank.EQ.0) THEN
         xd=1.0
         x=3.5
         print *,"process",world_rank," sends val  [3.50000000==]",x
         call head_d(x,xd,y,yd)
         print *,"process",world_rank," recvs val  [1.97095979==]",y
         print *,"process",world_rank," recvs diff [4.52341353==]",yd
         y=y+x
      ELSE IF (world_rank.EQ.1) THEN
         call head_d(x,xd,y,yd)
      ENDIF
      call AMPI_Finalize_NT(err_code)
      END

File: AdjoinableMPIexamples/Tapenade_C/IsendRecvWaitTwoWay/programAMPI_d.c
/*  DESIRED TAPENADE TANGENT USING THE AMPI LIBRARY
   Simple Isend+Wait || Recv followed by Recv || Isend+Wait */

#include <stdio.h>
#include <math.h>
#include "ampi/ampi.h"

void head_d(double *x, double *xd, double *y, double *yd) {
  AMPI_Request r;
  int world_rank;
  AMPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
  if (world_rank == 0) {
    *xd=*xd*2;
    *x=*x*2;
    TLS_AMPI_Isend(x, xd, 1, AMPI_ADOUBLE, AMPI_ADOUBLE, 1, 0,
                   AMPI_TO_RECV, MPI_COMM_WORLD, &r);
    TLS_AMPI_Recv(y, yd, 1, AMPI_ADOUBLE, AMPI_ADOUBLE, 1, 0,
                  AMPI_FROM_ISEND_WAIT, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
    TLS_AMPI_Wait(&r,MPI_STATUS_IGNORE);
    *yd=*yd*3;
    *y=*y*3;
  } else if (world_rank == 1) {
    double local;
    double locald;
    TLS_AMPI_Recv(&local, &locald, 1, AMPI_ADOUBLE, AMPI_ADOUBLE, 0, 0,
              AMPI_FROM_ISEND_WAIT, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
    locald=cos(local)*locald ;
    local=sin(local);
    TLS_AMPI_Isend(&local, &locald, 1, AMPI_ADOUBLE, AMPI_ADOUBLE, 0, 0,
               AMPI_TO_RECV, MPI_COMM_WORLD,&r);
    TLS_AMPI_Wait(&r,MPI_STATUS_IGNORE);
  }
}

int main(int argc, char** argv) {
  AMPI_Init_NT(0,0);
  int world_rank;
  AMPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
  double x,y;
  double xd,yd;
  if (world_rank == 0) {
    xd=1.0;
    x=3.5;
    printf(__FILE__ ": process %i sends val [3.500000==]%f\n", world_rank, x);
    head_d(&x,&xd,&y,&yd);
    printf(__FILE__ ": process %i recvs val [1.970960==]%f\n", world_rank, y);
    printf(__FILE__ ": process %i recvs diff [4.523414==]%f\n", world_rank, yd);
    y=y+x ;
  }else if (world_rank == 1) {
    head_d(&x,&xd,&y,&yd);
  } 
  AMPI_Finalize_NT();
  return 0;
}
